File: docs/eos/examples/compliance-starter/consent.go
// consent.go — per-purpose, versioned, revocable consent. Go parallel of consent.mjs / consent.py.
// GDPR Art.6/7 (lawful basis; withdrawal as easy as giving); PIPL separate consent for
// sensitive PI / cross-border (per-purpose flags, never a bundled all-or-nothing checkbox).
//
// In-memory stub: swap `store` for your DB table
//
//	(subject_id, purpose, granted, basis, version, at) with the latest row per (subject, purpose).
package compliance

// ConsentRecord is the latest state for one (subject, purpose).
type ConsentRecord struct {
	SubjectID string
	Purpose   string
	Granted   bool
	Basis     string
	Version   string
	At        string
}

// GrantOption tweaks the record written by Grant (idiomatic Go stand-in for the JS options
// object / Python keyword args). Use WithBasis / WithVersion.
type GrantOption func(*ConsentRecord)

// WithBasis sets the lawful basis (default "consent").
func WithBasis(basis string) GrantOption { return func(r *ConsentRecord) { r.Basis = basis } }

// WithVersion sets the consent notice version (default "v1").
func WithVersion(version string) GrantOption { return func(r *ConsentRecord) { r.Version = version } }

// ConsentStore records and queries per-purpose consent. Build via NewConsentStore.
type ConsentStore struct {
	store map[string]ConsentRecord // `${subjectId}::${purpose}` -> latest record
	Log   *AuditLog
}

// NewConsentStore builds a store; pass nil to get a fresh audit trail.
func NewConsentStore(log *AuditLog) *ConsentStore {
	if log == nil {
		log = NewAuditLog()
	}
	return &ConsentStore{store: map[string]ConsentRecord{}, Log: log}
}

func consentKey(subjectID, purpose string) string { return subjectID + "::" + purpose }

// Grant records consent for ONE purpose. Separate purposes = separate consent (PIPL).
func (c *ConsentStore) Grant(subjectID, purpose string, opts ...GrantOption) ConsentRecord {
	rec := ConsentRecord{SubjectID: subjectID, Purpose: purpose, Granted: true, Basis: "consent", Version: "v1", At: now()}
	for _, o := range opts {
		o(&rec)
	}
	c.store[consentKey(subjectID, purpose)] = rec
	c.Log.Record(Entry{"type": "consent.grant", "subjectId": subjectID, "purpose": purpose, "version": rec.Version, "basis": rec.Basis})
	return rec
}

// Revoke withdraws consent — must be as easy as granting (GDPR Art.7(3)). No data logged.
func (c *ConsentStore) Revoke(subjectID, purpose string) ConsentRecord {
	rec := ConsentRecord{SubjectID: subjectID, Purpose: purpose, Granted: false, At: now()}
	c.store[consentKey(subjectID, purpose)] = rec
	c.Log.Record(Entry{"type": "consent.revoke", "subjectId": subjectID, "purpose": purpose})
	return rec
}

// Check gates a purpose-bound processing step. True only if the latest record grants it.
func (c *ConsentStore) Check(subjectID, purpose string) bool {
	rec, ok := c.store[consentKey(subjectID, purpose)]
	return ok && rec.Granted
}

// State returns the queryable consent state for a subject (feeds DSAR "access" + a dashboard).
func (c *ConsentStore) State(subjectID string) map[string]ConsentRecord {
	out := map[string]ConsentRecord{}
	for _, v := range c.store {
		if v.SubjectID == subjectID {
			out[v.Purpose] = v
		}
	}
	return out
}

File: docs/eos/examples/compliance-starter/Consent.java
// Consent.java — per-purpose, versioned, revocable consent. Java parallel of consent.mjs / consent.py.
// GDPR Art.6/7 (lawful basis; withdrawal as easy as giving); PIPL separate consent for
// sensitive PI / cross-border (per-purpose flags, never a bundled all-or-nothing checkbox).
//
// In-memory stub: swap `store` for your DB table
//   (subject_id, purpose, granted, basis, version, at) with the latest row per (subject, purpose).
import java.time.Instant;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

public final class Consent {
    private final Map<String, Map<String, Object>> store = new HashMap<>(); // "subject::purpose" -> latest record
    public final Audit log;

    public Consent(Audit log) {
        this.log = (log != null) ? log : Audit.create();
    }

    public static Consent create() {
        return new Consent(null);
    }

    public static Consent create(Audit log) {
        return new Consent(log);
    }

    private static String key(String subjectId, String purpose) {
        return subjectId + "::" + purpose;
    }

    private static String now() {
        return Instant.now().toString();
    }

    /** Record consent for ONE purpose with defaults (basis "consent", version "v1"). */
    public Map<String, Object> grant(String subjectId, String purpose) {
        return grant(subjectId, purpose, "consent", "v1");
    }

    /** Record consent for ONE purpose. Separate purposes = separate consent (PIPL). */
    public Map<String, Object> grant(String subjectId, String purpose, String basis, String version) {
        Map<String, Object> rec = new LinkedHashMap<>();
        rec.put("subjectId", subjectId);
        rec.put("purpose", purpose);
        rec.put("granted", true);
        rec.put("basis", basis);
        rec.put("version", version);
        rec.put("at", now());
        store.put(key(subjectId, purpose), rec);
        Map<String, Object> ev = new LinkedHashMap<>();
        ev.put("type", "consent.grant");
        ev.put("subjectId", subjectId);
        ev.put("purpose", purpose);
        ev.put("version", version);
        ev.put("basis", basis);
        log.record(ev);
        return rec;
    }

    /** Withdrawal must be as easy as granting (GDPR Art.7(3)). No data logged, just the action. */
    public Map<String, Object> revoke(String subjectId, String purpose) {
        Map<String, Object> rec = new LinkedHashMap<>();
        rec.put("subjectId", subjectId);
        rec.put("purpose", purpose);
        rec.put("granted", false);
        rec.put("at", now());
        store.put(key(subjectId, purpose), rec);
        Map<String, Object> ev = new LinkedHashMap<>();
        ev.put("type", "consent.revoke");
        ev.put("subjectId", subjectId);
        ev.put("purpose", purpose);
        log.record(ev);
        return rec;
    }

    /** Gate a purpose-bound processing step on this. */
    public boolean check(String subjectId, String purpose) {
        Map<String, Object> rec = store.get(key(subjectId, purpose));
        return rec != null && Boolean.TRUE.equals(rec.get("granted"));
    }

    /** Queryable consent state for a subject (feeds DSAR "access" + your consent dashboard). */
    public Map<String, Map<String, Object>> state(String subjectId) {
        Map<String, Map<String, Object>> out = new LinkedHashMap<>();
        for (Map<String, Object> rec : store.values()) {
            if (subjectId.equals(rec.get("subjectId"))) {
                out.put((String) rec.get("purpose"), rec);
            }
        }
        return out;
    }
}
